lib/Minion/Backend/MongoDB.pm

```diff
@@ -215,49 +477,99 @@ sub _notifications {
   return if grep { $_ eq $notifications->name } $self->mongodb->collection_names;
 
   $self->mongodb->run_command([create => $notifications->name, capped => 1, size => 1048576, max => 128]);
-  $notifications->insert({});
+  $notifications->insert_one({});
+}
+
+sub _total {
+    my ($name, $res, $tot) = @_;
+    return { total => $tot, $name => $res};
 }
 
 sub _try {
-  my ($self, $oid) = @_;
+  my ($self, $oid, $options) = @_;
 
-  my $doc = {
-    query => Tie::IxHash->new(
-      delayed => {'$lt' => DateTime->from_epoch(epoch => time)},
-      state   => 'inactive',
-      task => {'$in' => [keys %{$self->minion->tasks}]}
-    ),
-    fields => {args     => 1, task => 1},
-    sort   => {priority => -1},
-    update => {'$set' => {started => DateTime->from_epoch(epoch => time), state => 'active', worker => $oid}},
-    new    => 1
-  };
+  my $match = Tie::IxHash->new(
+    delayed   => {'$lt' => DateTime->from_epoch(epoch => time)},
+    state     => 'inactive',
+    task      => {'$in' => [keys %{$self->minion->tasks}]},
+    queue     => {'$in' => $options->{queues} // ['default']}
+  );
+  $match->Push('_id' => $options->{id}) if defined $options->{id};
+
+  my $docs = $self->jobs->find($match)->sort({priority => -1, id => 1});
+
+  my $find = 0;
+  my $doc_matched;
+  while ((my $doc = $docs->next) && !$find) {
+    # parents not exits or
+    # exists but are not in inactive, active, failed states
+    $find = !scalar @{$doc->{parents}} ||
+        !$self->jobs->count_documents({
+            _id => {'$in' => $doc->{parents}},
+            state => {'$in' => ['inactive', 'active', 'failed']}
+    });
+    $doc_matched = $doc if $find;
+  }
 
-  return $self->jobs->find_and_modify($doc);
+  return undef unless $doc_matched;
+
+  my $doc = [
+    {_id => $doc_matched->{_id}},
+    {'$set' => {
+        started => DateTime->from_epoch(epoch => time),
+        state => 'active',
+        worker => $oid
+    }},
+    {
+        projection      => {args => 1, retries => 1, task => 1},
+        upsert          => 0,
+        returnDocument  => 'after',
+    }
+  ];
+
+  my $job = $self->jobs->find_one_and_update(@$doc);
```

Comment:
This code is also prone to race condition. Several workers can gets one job.

Reply:
Same as lock. It seems there is possibility, if you pass the correct id in the options to get the same job of another worker

Reply:
In this case we can fix race condition by adding `state => 'inactive'` to the `_id => $doc_matched->{_id}` in filter `find_one_and_update`.